### ISSUE TITLE
x/y_range not properly set with tile source

### ISSUE THREAD

**Issue by lgasnier** [bokeh/bokeh#6655]
Initial extent can't be set as expected by setting ranges as explained in http://bokeh.pydata.org/en/latest/docs/user_guide/plotting.html#setting-ranges.

However, extent can be well adjust manually with Wheel Zoom and Box Zoom.

See script below.
If we comment the two lines :
```
wms = BBoxTileSource(url=url)
p.add_tile(wms)
```
initial extent is properly set.

Note that url called by BBoxTileSource is a not tiled web map service (WMS getmap()).
The same behavior occurs with Tiled WMS (called with WMTSTileSource).

Code example:
Map extent should be equal to the colored square.

```
from bokeh.plotting import figure
from bokeh.models.tiles import BBoxTileSource, WMTSTileSource
from bokeh.io import show
from bokeh.io import output_file

output_file("test_getmap.html")

url = ("http://ws.carmencarto.fr/WMS/74/smartphone_mdo_nationale2016?"
        "&REQUEST=getmap&layers=mdo&BBOX={XMIN},{YMIN},{XMAX},{YMAX}"
        "&SRS=epsg:3857&FORMAT=image/png&WIDTH=700&HEIGHT=700")

xmin = 200000
xmax = 200000 + 1000000
ymin = 5000000
ymax = 5000000 + 1000000

p = figure(width=700, height=700, x_range=(200000, 200000 + 1000000),
           y_range=(5000000, 5000000 + 1000000))
p.axis.visible = False
#wmts = WMTSTileSource(url="http://c.tile.openstreetmap.org/{Z}/{X}/{Y}.png")
#p.add_tile(wmts)
wms = BBoxTileSource(url=url)
p.add_tile(wms)
p.quad(top=[ymax], bottom=[ymin], left=[xmin],
       right=[xmax], color="#B3DE69", alpha=0.5)
show(p)
```

**Comment by nanodan** [comment]
Is there a current workaround for this?

**Comment by lgasnier** [comment]
Sorry, I didn't find a workaround.

**Comment by TiemenSch** [comment]
Perhaps this is more broad than the tile? 

I'm having issues with a plot to which I add two sets of rects using p.rect().

At first, the axis behave as normal and leave a little whitespace around all rects. 

After the second one though, the whitespace on the beginning of the axis becomes way too much.

This only happens when the rects are somewhat "near" the (0,0) value, but still farther away than the initial set. This would leave me to believe that the whitespace need not increase. 

Are the axis (or range!) bounds defined differently when additional renderers are added?

**Comment by bryevdv** [comment]
@TiemenSch I don't think what you describe is directly related to the original issue, which I judge to be a result of specific things related to tile renderers. However it's hard for me to envisage exactly what's happening from your description. Actual code and a screen capture would go a long way to helping us diagnose if there is a problem, but please make a new separate issue.

**Comment by bryevdv** [comment]
@lgasnier @nanodan I expect the problem is here:

https://github.com/bokeh/bokeh/blob/master/bokehjs/src/coffee/models/tiles/tile_renderer.coffee#L60-L68

I will try to look at this in the near future. I'm not really sure why the renderer should be setting range values itself there. I guess it seems to be so that the range starts out exactly at one of the defined zoom levels. But I don't see why that should be required, or at least it should be configurable. So the solution might just be to delete that block of code entirely, or add a property to turn that behavior off. 

@brendancol do you have any additional comments?

**Comment by lgasnier** [comment]
I think aspect ratio must be kept for tile renderers to always preserve map aspect.

However, is there a way to change x/y_range after L63 in *tile_renderer.coffee* to take into account user values with something like https://github.com/bokeh/bokeh/blob/master/bokehjs/src/coffee/models/plots/plot_canvas.coffee#L217-L231?

I don't known exactly how to do that in *tile_renderer.coffee* but I simulated it with a button and javascript in the code below.

```
from bokeh.plotting import figure
from bokeh.models.tiles import WMTSTileSource
from bokeh.io import show
from bokeh.io import output_file
from bokeh.models import Button
from bokeh.models import CustomJS, Button
from bokeh.models import BoxZoomTool
from bokeh.layouts import widgetbox, column

output_file('test_xyrange_WMTS.html')

width = 700
height = 500

xmin = 200000
xmax = 200000 + 1000000
ymin = 5000000
ymax = 5000000 + 1000000

p = figure(width=width, height=height, 
           x_range=(xmin, xmax),
           y_range=(ymin, ymax),
           tools="pan,wheel_zoom,reset")

p.add_tools(BoxZoomTool(match_aspect=True))

wmts = WMTSTileSource(url="http://c.tile.openstreetmap.org/{Z}/{X}/{Y}.png")
p.add_tile(wmts)
p.quad(top=[ymax], bottom=[ymin], left=[xmin],
       right=[xmax], color="#B3DE69", alpha=0.5)

code = """
    var r = fig.inner_width / fig.inner_height;
    var xmin = 200000;
    var xmax = 200000 + 1000000;
    var ymin = 5000000;
    var ymax = 5000000 + 1000000;
    
    var width = Math.abs(xmax - xmin);
    var height = Math.abs(ymax - ymin);
    var xcenter = 0.5*(xmax + xmin);
    var ycenter = 0.5*(ymax + ymin);
    if (width < r*height) {
        width = r*height;
    }
    else {
        height = width/r;
    }
    fig.x_range.start = xcenter - width/2;
    fig.x_range.end = xcenter + width/2;
    fig.y_range.start = ycenter - height/2;
    fig.y_range.end = ycenter + height/2;
    
"""
cb = CustomJS(args={'fig': p},
               code=code)

button = Button(label="Foo", button_type="success")
button.js_on_click(cb)
p2 = column(button, p)
show(p2)
```
I hope it can be useful for you.
Thanks.

**Comment by philippjfr** [comment]
It would helpful to know if my change in #7294 satisfactorily addresses everyone's issue, so please chime in if the solution works for you or you'd expect some different behavior entirely.

**Comment by bryevdv** [comment]
Going to consider this resolved by #7294  Please open a new issue if there are further or different requirements to consider
